src/name_list.c

```diff
@@ -0,0 +1,222 @@
+
+static bool is_compatible(enum name_flavor a, enum name_flavor b)
+{
+	if (b == NAME_UNKNOWN) {
+		return true;
+	}
+
+	switch (a) {
+	case NAME_TYPE:
+		return b == NAME_TYPE || b == NAME_TYPE_OR_ATTRIBUTE;
+	case NAME_TYPEATTRIBUTE:
+		return b == NAME_TYPEATTRIBUTE || b == NAME_TYPE_OR_ATTRIBUTE;
+	case NAME_TYPE_OR_ATTRIBUTE:
+		return b == NAME_TYPE || b == NAME_TYPEATTRIBUTE || b == NAME_TYPE_OR_ATTRIBUTE;
+	case NAME_ROLE:
+		return b == NAME_ROLE || b == NAME_ROLE_OR_ATTRIBUTE;
+	case NAME_ROLEATTRIBUTE:
+		return b == NAME_ROLEATTRIBUTE || b == NAME_ROLE_OR_ATTRIBUTE;
+	case NAME_ROLE_OR_ATTRIBUTE:
+		return b == NAME_ROLE || b == NAME_ROLEATTRIBUTE || b == NAME_ROLE_OR_ATTRIBUTE;
+	case NAME_CLASS:
+		return b == NAME_CLASS;
+	case NAME_PERM:
+		return b == NAME_PERM;
+	case NAME_USER:
+		return b == NAME_USER;
+	case NAME_BOOL:
+		return b == NAME_BOOL;
+	case NAME_UNKNOWN:
+		return true;
+	default:
+		// should never happen
+		return 0;
+	}
+}
+
+bool name_is_type(const struct name_data *name)
+{
+	return is_compatible(name->flavor, NAME_TYPE);
+}
+
+bool name_is_typeattr(const struct name_data *name)
+{
+	return is_compatible(name->flavor, NAME_TYPEATTRIBUTE);
+}
+
+bool name_is_role(const struct name_data *name)
+{
+	return is_compatible(name->flavor, NAME_ROLE);
+}
+
+bool name_is_roleattr(const struct name_data *name)
+{
+	return is_compatible(name->flavor, NAME_ROLEATTRIBUTE);
+}
+
+bool name_is_class(const struct name_data *name)
+{
+	return is_compatible(name->flavor, NAME_CLASS);
+}
+
+bool name_list_contains_name(const struct name_list *nl, const struct name_data *name)
+{
+	for (;nl; nl = nl->next) {
+		if (!is_compatible(nl->data->flavor, name->flavor)) {
+			continue;
+		}
+
+		if (0 == strcmp(nl->data->name, name->name)) {
+			return true;
+		}
+	}
+	return false;
+}
+
+struct name_list *name_list_from_sl_with_traits(const struct string_list *sl,
+                                                enum name_flavor flavor,
+                                                const struct string_list *traits)
+{
+	if (!sl) {
+		return NULL;
+	}
+	struct name_list *ret = xmalloc(sizeof(struct name_list));
+	struct name_list *cur = ret;
+
+	while (sl) {
+		struct name_data *data = xmalloc(sizeof(struct name_data));
+		data->name = xstrdup(sl->string);
+		data->flavor = flavor;
+		data->traits = copy_string_list(traits);
+		cur->data = data;
+
+		if (sl->next) {
+			cur->next = xmalloc(sizeof(struct name_list));
+		} else {
+			cur->next = NULL;
+		}
+		sl = sl->next;
+		cur = cur->next;
+	}
+	return ret;
+}
+
+struct name_list *name_list_create(const char *name, enum name_flavor flavor)
+{
+	struct name_data *data = xmalloc(sizeof(struct name_data));
+	data->name = xstrdup(name);
+	data->flavor = flavor;
+	data->traits = NULL;
+	struct name_list *ret = xmalloc(sizeof(struct name_list));
+	ret->data = data;
+	ret->next = NULL;
+
+	return ret;
+}
+
+struct name_list *name_list_from_decl(const struct declaration_data *decl)
```

Comment:
Since commit https://github.com/SELinuxProject/selint/commit/3f83758c1825c1e720867e18ab97ee18ef0eadf9 each declaration of the same statement in a require block gets parsed into its own node.